coretrace/simulation_data/cst_templates/heliostat.cpp

```diff
@@ -0,0 +1,437 @@
+      canting_method(UNSET),
+      onaxis_canting_distance(-1.0),
+      offaxis_canting_sun_position_azimuth(-1.0),
+      offaxis_canting_sun_position_zenith(-1.0),
+      heliostat_area(-1.0),
+      tracking_azimuth(-1.0),
+      tracking_elevation(-1.0),
+      target_set(false)
+{
+    this->elevation_axis.set_values(1.0, 0.0, 0.0);
+    this->sun_position.set_values(0.0, 0.0, 1.0);
+    this->optics_mirror.set_ideal_reflection();
+    return;
+}
+
+Heliostat::~Heliostat()
+{
+    this->facets.clear();
+    return;
+}
+
+void Heliostat::create_geometry()
+{
+    if (this->initialized)
+        return;
+
+    this->enforce_user_fields_set();
+
+    this->facets.clear();
+    this->clear();
+
+    double panel_len_x = this->aperture_size_x -
+                         this->gap_x * (this->num_panels_x - 1);
+    panel_len_x /= this->num_panels_x;
+    double panel_len_y = this->aperture_size_y -
+                         this->gap_y * (this->num_panels_y - 1);
+    panel_len_y /= this->num_panels_y;
+
+    if (this->canting_method == OFF_AXIS)
+    {
+        // TODO: Do stuff here...
+    }
+
+    this->heliostat_area = 0.0;
+    double panel_x = -0.5 * (this->aperture_size_x - panel_len_x);
+    double panel_y = -0.5 * (this->aperture_size_y - panel_len_y);
+
+    single_element_ptr elem;
+    Vector3d origin;
+    Vector3d aim;
+    aperture_ptr ap;
+    surface_ptr surf;
+    element_id sts;
+    double c, z;
+
+    for (int j = 0; j < this->num_panels_y; ++j)
+    {
+        for (int i = 0; i < this->num_panels_x; ++i)
+        {
+            elem = make_element<SingleElement>();
+
+            if (this->canting_method == NONE)
+            {
+                origin.set_values(panel_x, panel_y, 0.0);
+                aim.set_values(panel_x, panel_y, 1000.0);
+            }
+            else if (this->canting_method == ON_AXIS)
+            {
+                c = 0.5 / (this->onaxis_canting_distance);
+                z = 0.5 * c * (panel_x * panel_x + panel_y * panel_y);
+                origin.set_values(panel_x, panel_y, z);
+                aim.set_values(0.0, 0.0, 2.0 * this->onaxis_canting_distance);
+            }
+            else if (this->canting_method == OFF_AXIS)
+            {
+                origin.set_values(panel_x, panel_y, 0.0);
+                // TODO: Set aim vector values
+                aim.set_values(0.0, 0.0, 1.0);
+                throw std::runtime_error("OFF_AXIS is not yet implemented");
+            }
+            else if (this->canting_method == UNSET)
+            {
+                throw std::runtime_error("Canting method unset");
+            }
+            else
+            {
+                throw std::runtime_error("Unknown canting method");
+            }
+
+            ap = make_aperture<Rectangle>(panel_len_x, panel_len_y);
+            elem->set_aperture(ap);
+
+            // // TODO: Is this the correct thing to do here?
+            // vector_add(1.0, this->focal_point, -1.0, origin, aim);
```

Comment:
What is the focal_point? Should this and the other code referring to it be removed?

Reply:
Probably something from when I was trying to implement it. I will remove it.